File: Lineup Builder/AutoFillCoordinator.swift
import Combine
import Foundation

// MARK: - AutoFillCoordinator
//
// One place that knows how to run an Auto-Fill end to end: parse the coach's
// natural-language prompt (racing a timeout), call AutoFillEngine, compose the
// coach-facing messages, and emit analytics.
//
// Three callers need that sequence — the iPhone grid, the iPad summary pane,
// and FillLineupIntent. Before this existed the first two each had their own
// copy, and they had already drifted: the iPhone version rebuilt the
// "couldn't fill" text in a private helper that duplicated
// AutoFillResult.incompleteMessage(multiInning:) almost line for line, the two
// alert-sequencing delays differed, and only one of them captured an undo
// snapshot. Adding a third copy for the intent is how that becomes permanent.
//
// What stays with the callers: alerts, toasts, popover state, undo. This type
// deliberately owns no UI state and never touches LineupStore — it takes a
// Team in and hands an outcome back, which is also what lets an App Intent use
// it without a view hierarchy.

/// What a fill should cover. Inning indices are zero-based, matching
/// `Lineup.innings`, and only become 1-based in coach-facing text.
nonisolated enum AutoFillScope: Equatable {
    /// Just the inning currently on screen.
    case inning(Int)
    /// Innings 1 through this index, inclusive.
    case through(Int)

    var isMultiInning: Bool {
        if case .through = self { return true }
        return false
    }

    /// How this scope reads in a sentence: "inning 3", "innings 1–6".
    var label: String {
        switch self {
        case .inning(let index):  return "inning \(index + 1)"
        case .through(let last):  return "innings 1–\(last + 1)"
        }
    }

    /// Analytics dimension. Matches the values the shipped `autofill.used`
    /// signal already reports, so the existing funnel stays continuous.
    var analyticsMode: String {
        switch self {
        case .inning:  return "inning"
        case .through: return "range"
        }
    }
}

/// Everything one Auto-Fill produced, with no UI attached.
nonisolated struct AutoFillOutcome {
    let scope: AutoFillScope
    /// The filled lineup. Only worth writing back when `didFill` is true.
    let lineup: Lineup
    let filledCount: Int
    let unfilledSlots: [AutoFillUnfilledSlot]
    /// Why some slots stayed empty, or nil when everything filled.
    let incompleteMessage: String?
    /// Parse misses and engine overrides, already combined into one block of
    /// text, or nil when there's nothing to report.
    let noticeMessage: String?

    var didFill: Bool { filledCount > 0 }
    var hasUnfilledSlots: Bool { !unfilledSlots.isEmpty }

    /// Undo toast copy: "Auto-filled 9 positions (inning 3)".
    var undoMessage: String {
        let noun = filledCount == 1 ? "position" : "positions"
        return "Auto-filled \(filledCount) \(noun) (\(scope.label))"
    }

    /// What Siri says back. The alert copy is written for a screen the coach is
    /// already looking at, so the lead sentence here has to carry the result on
    /// its own before any caveats follow.
    var spokenSummary: String {
        var lead: String
        if didFill {
            let noun = filledCount == 1 ? "position" : "positions"
            lead = "Filled \(filledCount) \(noun) in \(scope.label)."
        } else if hasUnfilledSlots {
            lead = "I couldn't fill anything in \(scope.label)."
        } else {
            lead = "Every position in \(scope.label) was already assigned."
        }

        for detail in [incompleteMessage, noticeMessage].compactMap({ $0 }) {
            lead += " " + detail.replacingOccurrences(of: "\n\n", with: " ")
                                .replacingOccurrences(of: "\n", with: " ")
        }
        return lead
    }
}

@MainActor
final class AutoFillCoordinator: ObservableObject {

    /// True while the on-device parse is in flight. Drives the popover's
    /// "Reading your instructions…" spinner.
    @Published private(set) var isParsingPrompt = false

    /// Owns the on-device LanguageModelSession. Rebuilt on every prewarm so the
    /// roster and inning count baked into the session are always current.
    private var nlService: AutoFillNLConstraintService?

    init() {}

    // MARK: - NL Session Lifecycle

    /// Builds a fresh NL service for this team and warms the on-device model.
    /// Call when the Auto-Fill popover opens, **not** when Fill is tapped.
    ///
    /// This is the single biggest speed lever in the feature. Creating the
    /// session inside `parse()` made every Fill tap pay full model load plus
    /// instruction prefill before a single token could be generated; doing it
    /// here spends that cost during the seconds the coach is already typing.
    func prewarm(for team: Team) {
        let service = makeService(for: team)
        service.prewarm()
        nlService = service
    }

    /// Popover closed without a fill. Drop the session so a warm model isn't
    /// held open indefinitely.
    func teardown() {
        nlService = nil
    }

    private func makeService(for team: Team) -> AutoFillNLConstraintService {
        AutoFillNLConstraintService(
            activePlayers: team.lineup.activePlayers(from: team.players),
            inningCount: team.lineup.innings.count
        )
    }

    // MARK: - Running a Fill

    /// Parses `prompt` (if any), runs AutoFillEngine over `team`, and composes
    /// the coach-facing messages.
    ///
    /// A blank prompt, or an on-device parse failure, falls back to
    /// unconstrained behavior rather than blocking the fill — a coach who just
    /// wants the normal Auto-Fill shouldn't be interrupted by an Apple
    /// Intelligence availability check they never asked to trigger.
    ///
    /// Callers are responsible for writing `outcome.lineup` back and for every
    /// piece of UI that follows.
    func run(scope: AutoFillScope, prompt: String, team: Team) async -> AutoFillOutcome {
        var constraints = AutoFillConstraintSet.empty
        var parseDiagnosticMessage: String? = nil

        let trimmedPrompt = prompt.trimmingCharacters(in: .whitespacesAndNewlines)
        if !trimmedPrompt.isEmpty {
            isParsingPrompt = true
            let parseResult = await parseWithTimeout(prompt: prompt, team: team)
            constraints = parseResult.constraints
            parseDiagnosticMessage = parseResult.diagnosticMessage
            Analytics.signal("autofill.nl_prompt_used", parameters: [
                "constraintCount": "\(constraints.playerConstraints.count)",
                "diagnosticCount": "\(parseResult.diagnostics.count)"
            ])
            isParsingPrompt = false

            // Safety net for game-wide pattern rules. The on-device model parse
            // can time out (8s cap) or throw — on hardware far more than in the
            // model-less simulator — and both cases return `.empty`, silently
            // dropping a rule the coach plainly typed. Pattern detection is pure,
            // instant text matching that never needed the model, so re-apply it
            // here over whatever the parse produced. Idempotent: a no-op when the
            // parse already set the flag.
            let before = constraints.patternRules
            (constraints, parseDiagnosticMessage) = Self.applyingPatternRuleSafetyNet(
                to: constraints, prompt: trimmedPrompt, diagnostic: parseDiagnosticMessage
            )
            if constraints.patternRules != before {
                Analytics.signal("autofill.pattern_rule_safety_net", parameters: [
                    "rule": "benchInConsecutivePairs"
                ])
            }
        }

        let preferences = Dictionary(
            uniqueKeysWithValues: team.players.map { ($0.id, $0.positionPreferences) }
        )

        let result: AutoFillResult
        switch scope {
        case .inning(let index):
            result = AutoFillEngine.fillInning(
                index,
                in: team.lineup,
                players: team.players,
                preferences: preferences,
                config: team.fairPlayConfig,
                pitchingConfig: team.pitchingConfig,
                gameLogs: team.gameLogs,
                constraints: constraints
            )
        case .through(let lastInning):
            result = AutoFillEngine.fillInnings(
                through: lastInning,
                in: team.lineup,
                players: team.players,
                preferences: preferences,
                config: team.fairPlayConfig,
                pitchingConfig: team.pitchingConfig,
                gameLogs: team.gameLogs,
                constraints: constraints
            )
        }

        Analytics.signal("autofill.used", parameters: [
            "mode": scope.analyticsMode,
            "filledCount": "\(result.filledCount)",
            "unfilledCount": "\(result.unfilledSlots.count)"
        ])

        return Self.outcome(
            from: result,
            scope: scope,
            players: team.players,
            parseDiagnosticMessage: parseDiagnosticMessage
        )
    }

    /// Message composition, split out so it can be tested without an on-device
    /// model or a live team.
    ///
    /// Parse-side diagnostics (an instruction we couldn't resolve at all) and
    /// engine-side notices (an instruction we honored by bending a fair-play
    /// guideline) are deliberately one message: from the coach's point of view
    /// they're all "here's what happened to what you asked for". Parse misses
    /// lead, since a skipped instruction is the more surprising outcome.
    nonisolated static func outcome(
        from result: AutoFillResult,
        scope: AutoFillScope,
        players: [Player],
        parseDiagnosticMessage: String?
    ) -> AutoFillOutcome {
        let combinedNotice = [parseDiagnosticMessage, result.constraintNoticeMessage(players: players)]
            .compactMap { $0 }
            .joined(separator: "\n")

        return AutoFillOutcome(
            scope: scope,
            lineup: result.lineup,
            filledCount: result.filledCount,
            unfilledSlots: result.unfilledSlots,
            incompleteMessage: result.incompleteMessage(multiInning: scope.isMultiInning),
            noticeMessage: combinedNotice.isEmpty ? nil : combinedNotice
        )
    }

    // MARK: - Pattern-Rule Safety Net
    //
    // Re-applies deterministic, model-free pattern detection over whatever the
    // parse produced. Pure and standalone so it can be unit-tested without an
    // on-device model — which is exactly the situation it exists to cover (the
    // model timed out or threw, so the parse came back empty). Idempotent: when
    // the parse already carried the rule, nothing changes.
    nonisolated static func applyingPatternRuleSafetyNet(
        to constraints: AutoFillConstraintSet,
        prompt: String,
        diagnostic: String?
    ) -> (AutoFillConstraintSet, String?) {
        let detected = AutoFillNLConstraintService.detectedPatternRules(in: prompt)
        guard detected.benchInConsecutivePairs,
              !constraints.patternRules.benchInConsecutivePairs else {
            return (constraints, diagnostic)
        }
        let merged = AutoFillConstraintSet(
            playerConstraints: constraints.playerConstraints,
            patternRules: AutoFillPatternRules(benchInConsecutivePairs: true)
        )
        let message = [diagnostic, AutoFillNLConstraintService.benchPairingConfirmationMessage]
            .compactMap { $0 }
            .joined(separator: "\n")
        return (merged, message)
    }

    // MARK: - Parse Timeout

    /// Races the NL parse against a timeout so a slow or stuck on-device model
    /// call can never leave the coach staring at a spinner indefinitely. Falls
    /// back to unconstrained after `timeout` elapses.
    ///
    /// With prewarming in place the session is already warm and the
    /// instructions already prefilled by the time this runs, so a normal parse
    /// should land well inside the timeout. The old 12s ceiling was sized for a
    /// cold start on every call; 8s is a more honest bound now, and if the model
    /// hasn't answered in 8 seconds with a warm session, waiting longer is
    /// unlikely to help.
    ///
    /// Note: if the underlying model call ignores cancellation it may keep
    /// running in the background after the timeout wins. That's harmless (the
    /// result is discarded), just not free.
    private func parseWithTimeout(
        prompt: String,
        team: Team,
        timeout: Duration = .seconds(8)
    ) async -> AutoFillNLParseResult {
        // prewarm() should have built this. If it somehow didn't — an intent
        // invocation never opens a popover — build one now rather than
        // silently skipping the parse and ignoring what the coach asked for.
        let service = nlService ?? {
            let fresh = makeService(for: team)
            nlService = fresh
            return fresh
        }()

        // Run the parse as its own main-actor task, then race its *handle*
        // against a timeout. Racing the Task (which is Sendable) rather than
        // capturing the non-Sendable `service` directly inside the task-group
        // closures is what keeps this legible to Swift 6's region-isolation
        // checker — a plain `group.addTask { @MainActor in service.parse(...) }`
        // trips a checker limitation on the captured main-actor service.
        let parseTask = Task { @MainActor in
            do {
                return try await service.parse(prompt: prompt)
            } catch {
                // Apple Intelligence unavailable, or the model call failed.
                // Proceed unconstrained rather than blocking the fill.
                return AutoFillNLParseResult.empty
            }
        }

        // `nil` is the timeout marker; a real parse result is always non-nil.
        return await withTaskGroup(of: AutoFillNLParseResult?.self) { group in
            group.addTask { await parseTask.value }
            group.addTask {
                try? await Task.sleep(for: timeout)
                return nil
            }
            let first = await group.next() ?? nil
            group.cancelAll()
            // If the timeout won, stop the parse. Harmless if the model call
            // ignores cancellation — its result is simply discarded.
            parseTask.cancel()
            return first ?? .empty
        }
    }
}

File: web/src/core/autofillCoordinator.ts
// Runs one Auto-Fill for the grid and turns the result into what the coach
// sees. Port of AutoFillCoordinator.run/outcome and AutoFillResult's
// incompleteMessage / constraintNoticeMessage; message wording is the iOS copy.
//
// Web differences:
//  * No on-device model. The deterministic parser always runs; when it can't
//    confidently read an instruction (iOS would ask Apple Intelligence), its
//    best-effort constraints are still used and the coach is told.
//  * The pitch-rules reference date is the lineup's game date (iOS Auto-Fill
//    used today's date; its grid warning already used the game date).
//  * The unfilled-pitcher copy mentions rest days, since the web also blocks
//    pitchers who must rest.

import { AutoFillParser, BENCH_PAIRING_CONFIRMATION, detectedPatternRules } from './autofillParser';
import { fill } from './autofillEngine';
import type { AutoFillResult, ConstraintSet, ConstraintTarget } from './autofillTypes';
import type { FairPlayConfig, GameLog, Lineup, PitchingConfig, Player } from './model';
import { shortName } from './lineupOps';

export type FillScope = { kind: 'inning'; inning: number } | { kind: 'through'; inning: number };

export interface AutoFillOutcome {
  scope: FillScope;
  lineup: Lineup;
  filledCount: number;
  /** Why some slots stayed empty, or null. */
  incompleteMessage: string | null;
  /** Parse misses and instruction notices, combined, or null. */
  noticeMessage: string | null;
  /** "Auto-filled 9 positions (inning 3)". */
  undoMessage: string;
}

export const scopeLabel = (s: FillScope) => (s.kind === 'inning' ? `inning ${s.inning + 1}` : `innings 1–${s.inning + 1}`);

export function runAutoFill(args: {
  scope: FillScope;
  prompt: string;
  lineup: Lineup;
  players: Player[];
  config: FairPlayConfig;
  pitchingConfig?: PitchingConfig;
  gameLogs: GameLog[];
  random?: () => number;
}): AutoFillOutcome {
  const { scope, lineup, players } = args;
  const prompt = args.prompt.trim();
  let constraints: ConstraintSet = { playerConstraints: [], patternRules: { benchInConsecutivePairs: false } };
  const parseNotes: string[] = [];

  if (prompt) {
    const absent = new Set(lineup.absentPlayerIDs);
    const active = players.filter((p) => !absent.has(p.id));
    const parse = new AutoFillParser(active, lineup.innings.length).parseDeterministically(prompt);
    constraints = { playerConstraints: parse.constraints, patternRules: { benchInConsecutivePairs: false } };
    if (parse.hasUnresolvedInstruction) {
      parseNotes.push(parse.constraints.length
        ? 'Part of your note wasn\'t clear enough to follow, so only the instructions that were clear were used. Check the lineup, and try one instruction per line, like "Jake pitches innings 1 to 2".'
        : 'Your note wasn\'t clear enough to follow, so the lineup was filled without it. Try one instruction per line, like "Jake pitches innings 1 to 2".');
    }
    // Pattern rules (no player named) come from the same text detection on every path.
    if (detectedPatternRules(prompt).benchInConsecutivePairs) {
      constraints = { ...constraints, patternRules: { benchInConsecutivePairs: true } };
      parseNotes.push(BENCH_PAIRING_CONFIRMATION);
    }
  }

  const result = fill({
    scope: scope.kind === 'inning' ? { kind: 'inning', inning: scope.inning } : { kind: 'through', inning: scope.inning },
    lineup, players, config: args.config, pitchingConfig: args.pitchingConfig, gameLogs: args.gameLogs,
    constraints, referenceDate: lineup.gameDate, random: args.random,
  });

  const notice = [parseNotes.join('\n') || null, constraintNoticeMessage(result, players)].filter(Boolean).join('\n');
  const noun = result.filledCount === 1 ? 'position' : 'positions';
  return {
    scope,
    lineup: result.lineup,
    filledCount: result.filledCount,
    incompleteMessage: incompleteMessage(result, scope.kind === 'through'),
    noticeMessage: notice || null,
    undoMessage: `Auto-filled ${result.filledCount} ${noun} (${scopeLabel(scope)})`,
  };
}

export function incompleteMessage(result: AutoFillResult, multiInning: boolean): string | null {
  const slots = result.unfilledSlots;
  if (!slots.length) return null;
  const label = (group: typeof slots) => {
    if (!multiInning) return group.map((s) => s.position).join(', ');
    const byPos = new Map<string, number[]>();
    for (const s of group) byPos.set(s.position, [...(byPos.get(s.position) ?? []), s.inningIndex + 1]);
    return [...byPos.entries()]
      .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
      .map(([pos, innings]) => {
        const nums = innings.sort((a, b) => a - b).join(', ');
        return `${pos} (${innings.length === 1 ? 'Inning' : 'Innings'} ${nums})`;
      })
      .join(', ');
  };
  const parts: string[] = [];
  const of = (reason: string) => slots.filter((s) => s.reason === reason);
  if (of('rosterTooSmall').length) {
    parts.push(`${label(of('rosterTooSmall'))} could not be filled. Not enough active players to cover every field position. Mark additional players as active in the batting order, or assign these slots manually.`);
  }
  if (of('pitcherReentry').length) {
    parts.push(`${label(of('pitcherReentry'))} could not be filled. All eligible pitchers have already left the mound this game and cannot re-enter. Assign pitcher manually.`);
  }
  if (of('pitchCapacityLimited').length) {
    parts.push(`${label(of('pitchCapacityLimited'))} could not be filled. Every eligible pitcher either needs rest under your pitch count rules or has reached their estimated pitch limit for this game. Assign pitcher manually or check pitch counts on the Pitching tab.`);
  }
  if (of('neverPreferences').length) {
    parts.push(`${label(of('neverPreferences'))} could not be filled. All remaining players have those positions set to Never in their preferences. Update preferences on the Roster tab, or assign manually.`);
  }
  return parts.join('\n\n') || null;
}

export function constraintNoticeMessage(result: AutoFillResult, players: Player[]): string | null {
  const { constraintOverrides: overrides, constraintRejections: rejections } = result;
  if (!overrides.length && !rejections.length) return null;
  const name = (id: string) => {
    const p = players.find((x) => x.id === id);
    return p ? shortName(p) : 'A player';
  };
  const targetLabel = (t: ConstraintTarget) =>
    t.kind === 'position' ? t.position : t.kind === 'infield' ? 'an infield position'
    : t.kind === 'outfield' ? 'an outfield position' : 'Bench';
  const inningList = (innings: number[]) => {
    const n = [...innings].sort((a, b) => a - b).map((i) => String(i + 1));
    if (n.length === 1) return `inning ${n[0]}`;
    if (n.length === 2) return `innings ${n[0]} and ${n[1]}`;
    return `innings ${n.slice(0, -1).join(', ')}, and ${n[n.length - 1]}`;
  };
  const byPlayer = (reason: string) => {
    const groups = new Map<string, number[]>();
    for (const o of overrides.filter((x) => x.reason === reason)) {
      groups.set(o.playerID, [...(groups.get(o.playerID) ?? []), o.inningIndex]);
    }
    return [...groups.entries()].sort(([a], [b]) => (a < b ? -1 : 1));
  };

  const lines: string[] = [];
  for (const [id, inns] of byPlayer('pitcherSoftCapBypassed')) {
    lines.push(`${name(id)} also pitched ${inningList(inns)} — past the usual 2-inning guideline.`);
  }
  for (const [id, inns] of byPlayer('pitcherSoftCapBypassedByFallback')) {
    lines.push(`${name(id)} ended up pitching ${inningList(inns)} since no one else was eligible — past the usual 2-inning guideline.`);
  }
  for (const [id, inns] of byPlayer('benchedOutOfTurn')) {
    lines.push(`${name(id)} sat out ${inningList(inns)} ahead of the normal rotation turn.`);
  }
  for (const o of overrides) {
    if (o.reason === 'fairPlayZoneSkipped.infield') lines.push(`${name(o.playerID)} still hasn't played infield this game.`);
    if (o.reason === 'fairPlayZoneSkipped.outfield') lines.push(`${name(o.playerID)} still hasn't played outfield this game.`);
  }
  for (const r of rejections) {
    lines.push(`Couldn't put ${name(r.playerID)} at ${targetLabel(r.target)} in ${inningList([r.inningIndex])} — ${r.reason}`);
  }
  if (!lines.length) return null;
  return lines.length > 1 ? lines.map((l) => `• ${l}`).join('\n') : lines[0];
}
